reasoning_gym/algorithmic/graph_color.py

```diff
@@ -0,0 +1,237 @@
+            return False, f"Adjacent vertices {u} and {v} both have color {color_u}."
+
+    return True, "The coloring is valid."
+
+
+def greedy_graph_coloring(puzzle):
+    """
+    Attempts to color the graph using a simple greedy algorithm.
+    (Note: This may fail if the graph requires more than the given number of colors.)
+
+    Args:
+        puzzle (dict): The puzzle specification.
+
+    Returns:
+        dict or None: A dictionary mapping vertices to colors if successful; otherwise, None.
+    """
+    vertices = puzzle["vertices"]
+    edges = puzzle["edges"]
+    color_options = puzzle["color_options"]
+
+    # Build an adjacency list for each vertex.
+    adjacency = {v: set() for v in vertices}
+    for u, v in edges:
+        adjacency[u].add(v)
+        adjacency[v].add(u)
+
+    coloring = {}
+    for v in vertices:
+        # Find colors already used by neighbors.
+        neighbor_colors = {coloring.get(neighbor) for neighbor in adjacency[v] if neighbor in coloring}
+        # Pick the first available color not used by any neighbor.
+        available = [color for color in color_options if color not in neighbor_colors]
+        if not available:
+            return None  # Failed to color with the given number of colors.
+        coloring[v] = available[0]
+    return coloring
+
+
+@dataclass
+class GraphColorConfig:
+    """Configuration for GraphColor puzzle generation"""
+
+    num_colors: int = 4
+    num_vertices: int = 10
+    edge_probability: float = 0.4
+    seed: Optional[int] = None
+    size: int = 500
+
+    def validate(self):
+        """Validate configuration parameters"""
+        assert self.edge_probability < 1, "edge_probability must be less than 1"
+
+
+class GraphColorDataset(ProceduralDataset):
+    """Generates graph coloring problems with configurable parameters"""
+
+    def __init__(self, config: GraphColorConfig):
+        super().__init__(config=config, seed=config.seed, size=config.size)
+
+    def __getitem__(self, idx: int) -> dict:
+        """Generate a single GraphColor task
+
+        Returns:
+            dict with keys:
+                - question: str, the task description
+                - answer: str, a solution string
+                - metadata: dict with generation parameters
+        """
+        rng = Random(self.seed + idx)
+
+        puzzle = None
+        solution = None
+        while solution is None:
+            puzzle = generate_graph_coloring_puzzle(
+                rng=rng,
+                num_vertices=self.config.num_vertices,
+                edge_probability=self.config.edge_probability,
+                num_colors=self.config.num_colors,
+            )
+            solution = greedy_graph_coloring(puzzle)
+
+        edges = str(puzzle["edges"])
+        question = f"""
+Please provide a coloring for this graph such that every vertex is not connected to a vertex of the same color. The graph has these properties:
+
+Edges: {edges}
+Vertices: {puzzle["vertices"]}
```

Comment:
To me it would feel a bit more natural with fist printing the Vertices and then the Edges.

Could you add a simple example to the string how the result should be formatted, e.g for a trivial graph?